### PR TITLE
Stop generating nondet pointees for function pointers

### PR DESCRIPTION
**Pull request by xbauch**
when building the entry point. Because that leads to declaring function-type
objects and then assigning values to them. Rather we should simply initialise
the function pointer with NONDET(fptr_type) and let the function-pointer removal
find the candidates.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e5d5c9faf2d577e3163a114d451aef206c1ee5d2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120202139

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 68c9554a3d1bd658a20969cb09775fa568a0fa75).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/144957668

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
This comes up a lot, maybe we should have a utility function like `is_pointer_to_object` or something of the sort?

File: regression/cbmc/Function_Pointer_Init/main.c
Comment (thk123) [inline regression/cbmc/Function_Pointer_Init/main.c]: :question: This comment doesn't make sense wrt to the test checking for verification failed?

File: src/ansi-c/c_nondet_symbol_factory.cpp
Comment (thk123) [inline src/ansi-c/c_nondet_symbol_factory.cpp]: :question: What happens with pointers to code then? Just random number? 
Read the PR description - that is worth a comment here explaining why we don't need to initialize it here

**Review by thk123 — COMMENTED** [review]
:no_entry_sign: Please add a test for when there _is_ a function of the right type, to check it is selected. And two functions of the right type, to check it picks between them (as I believe the new behaviour should / old behaviour did not).

File: regression/cbmc/Function_Pointer_Init/main.c
Comment (xbauch) [reply regression/cbmc/Function_Pointer_Init/main.c]: Extended.

File: src/ansi-c/c_nondet_symbol_factory.cpp
Comment (xbauch) [reply src/ansi-c/c_nondet_symbol_factory.cpp]: Now the symbol is explicitly assigned a nondet (of the function-ptr type).

**Comment by xbauch** [comment]
>  Please add a test for when there _is_ a function of the right type, to check it is selected. And two functions of the right type, to check it picks between them (as I believe the new behaviour should / old behaviour did not).

I've added the tests. Note that existing functions will only be considered as candidates if dereferenced somewhere in the code.

File: src/ansi-c/c_nondet_symbol_factory.cpp
Comment (thk123) [inline src/ansi-c/c_nondet_symbol_factory.cpp:62]: :pick: I'd still like a comment to the effect 
```
// leave as nondet_ptr to allow `remove_function_pointers`
// to replace the pointer. 
```

**Review by thk123 — APPROVED** [review]
LGTM, thanks for the extra tests. 
~This just needs code owner approval @tautschnig @kroening~

**Comment by thk123** [comment]
currently failing CI, but looks transient, have restarted the job

File: src/ansi-c/c_nondet_symbol_factory.cpp
Comment (xbauch) [reply src/ansi-c/c_nondet_symbol_factory.cpp:62]: Added.

**Comment by thk123** [comment]
This is now passing CI and just needs code owner approval @tautschnig @kroening

**Review by tautschnig — APPROVED** [review]

**Pull request merged**
